src/services/healthcareProfessionalService.ts

```diff
@@ -0,0 +1,1033 @@
+        }
+
+        if (Array.isArray(filters.ids) && filters.ids.length > 0) {
+            const looksInvalid = filters.ids.some(id => !/^[0-9a-fA-F-]{36}$/.test(id))
+
+            if (looksInvalid) {
+                return { data: [], hasErrors: false }
+            }
+        }
+
+        const limit = filters.limit ?? 20
+        const offset = filters.offset ?? 0
+
+        const supabase = getSupabaseClient()
+
+        // Start base query on hps table and apply JSONB filters via helper
+        let hpSelect = applyHpFilters(supabase.from('hps').select('*'), filters)
+
+        if (Array.isArray(filters.ids)) {
+            if (filters.ids.length === 0) {
+                // user explicitly asked for "no ids" → return empty list right away
+                return { data: [], hasErrors: false }
+            }
+
+            // only return the professionals whose id is in filters.ids
+            hpSelect = hpSelect.in('id', filters.ids)
+        }
+
+        // Fallback to createdDate DESC.
+        const orderBy = filters.orderBy?.[0]
+
+        if (orderBy?.fieldToOrder) {
+            hpSelect = hpSelect.order(orderBy.fieldToOrder, {
+                ascending: orderBy.orderDirection !== 'desc'
+            })
+        } else {
+            hpSelect = hpSelect.order('createdDate', { ascending: false })
+        }
+
+        /**
+         * Perform paged fetch after all filters are applied so the page is computed
+         * on the correctly filtered set — this matches test expectations.
+         */
+        const { data: hpRows, error: hpRowsError } = await hpSelect.range(
+            offset,
+            offset + limit - 1
+        )
+
+        if (hpRowsError) {
+            throw hpRowsError
+        }
+
+        /// If there’s no data on this page, return early
+        if (!hpRows?.length) {
+            return { data: [], hasErrors: false }
+        }
+
+        // Extract IDs from this page to batch-load relations from the junction table
+        const hpIds = hpRows.map(row => row.id as string)
+
+        if (hpIds.length === 0) {
+            // extremely defensive; should not happen, but keeps the function safe
+            const list = (hpRows as dbSchema.DbHealthcareProfessionalRow[])
+                .map(hp => mapDbHpToGql(hp, []))
+
+            return { data: list, hasErrors: false }
+        }
+
+        // Load facilities relations for these HPs to avoid N+1 queries
+        const { data: facilityRelationsForHPs, error: facilityRelationsForHPsError} = await supabase
+            .from('hps_facilities')
+            .select('hps_id, facilities_id')
+            .in('hps_id', hpIds)
+
+        if (facilityRelationsForHPsError) {
+            throw facilityRelationsForHPsError
+        }
+
+        // Build a lookup map: hpId → [facilityId, ...]
+        const facilityIdsByHpId = new Map<string, string[]>()
+
+        for (const relation of facilityRelationsForHPs ?? []) {
```

Comment:
can do a map() function here instead. (for loops instead of map() are a bad habit AI has)